File: skills/fixture/scripts/codegen-validators.py
#!/usr/bin/env python3
import os
import subprocess
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SCHEMA_DIR = ROOT / "schemas"
OUT_DIR = ROOT / "generated"
JTD_CODEGEN = os.environ.get("JTD_CODEGEN", "jtd-codegen")


def export_name(stem: str) -> str:
    return "validate_" + stem.replace("-", "_")


def main() -> None:
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    barrel_lines: list[str] = []

    for schema_path in sorted(SCHEMA_DIR.glob("*.jdt.json")):
        stem = schema_path.name.removesuffix(".jdt.json")
        snake = stem.replace("-", "_")
        out_path = OUT_DIR / f"{snake}.py"
        code = subprocess.check_output(
            [JTD_CODEGEN, "--target", "python", str(schema_path)],
            text=True,
        )
        out_path.write_text(code)
        barrel_lines.append(f"from .{snake} import validate as {export_name(stem)}")
        print(f"generated {out_path.relative_to(ROOT)}")

    (OUT_DIR / "__init__.py").write_text("\n".join(barrel_lines) + "\n")
    print("generated __init__.py barrel")


if __name__ == "__main__":
    main()

File: skills/fixture/scripts/codegen-validators.mjs
import { execFileSync } from "node:child_process";
import { mkdir, readdir, writeFile } from "node:fs/promises";
import path from "node:path";

const ROOT = path.resolve(import.meta.dirname, "..");
const SCHEMA_DIR = path.join(ROOT, "schemas");
const OUT_DIR = path.join(ROOT, "generated");
const JTD_CODEGEN = process.env.JTD_CODEGEN ?? "jtd-codegen";

function exportName(stem) {
  return "validate" + stem.replace(/(^|-)([a-z])/g, (_, __, c) => c.toUpperCase());
}

const schemas = (await readdir(SCHEMA_DIR)).filter((f) => f.endsWith(".jdt.json"));
await mkdir(OUT_DIR, { recursive: true });

const barrelLines = [];
for (const file of schemas.sort()) {
  const stem = file.replace(/\.jdt\.json$/, "");
  const schemaPath = path.join(SCHEMA_DIR, file);
  const outPath = path.join(OUT_DIR, `${stem}.mjs`);
  const code = execFileSync(JTD_CODEGEN, ["--target", "js", schemaPath], {
    encoding: "utf8",
  });
  await writeFile(outPath, code);
  barrelLines.push(`export { validate as ${exportName(stem)} } from "./${stem}.mjs";`);
  console.log(`generated ${path.relative(ROOT, outPath)}`);
}

await writeFile(path.join(OUT_DIR, "validators.mjs"), barrelLines.join("\n") + "\n");
console.log("generated validators.mjs barrel");
